sema for conditional conformances.

huonw opened the pull request:
Currently just removes restriction and adds some basic recording of conditionality, and feeds that information through to the GSB and constraint solver and anywhere else that needs it. This doesn't do the SILGen/IRGen/Runtime work required to be able to actually interact with these conformances at runtime.

huonw committed:
[Sema] Remove restriction on conditional conformances.

huonw committed:
[AST] NFC: factor out getting a SpecializedProtocolConformance's substitution map.

huonw committed:
[AST] Compute conditional requirements in a conformance.

This allows determining which requirements make a conformance conditional; as
in, which requirements aren't known as part of the type itself.

Additionally, use this to assert that a few builtin protocols aren't
conditionally-conformed-to, something we won't support for now.

huonw committed:
[AST] Test conditional requirement selection by having -debug-generic-signatures include conformances.

And make conformances' dump include the conditional requirements too.

huonw committed:
[GSB] Include any necessary conditional requirements.

This enabled requirement inference from conditional requirements in a many
cases (but not all).

huonw committed:
[test] Add order-dependency conditional conformance edge case to test suite.

huonw committed:
[Sema/CS] Convenience function for adding a Requirement as a constraint.

huonw committed:
[Sema/CS] Thread conditional requirements through specialization checking.

huonw committed:
[AST&Sema] Add assertions for conditional conformances.

These are just places that may need updating for conditional
conformances, and so having them flag eagerly will help in future.

huonw committed:
[Sema] Add conditional requirements as CS constraints.

This tries to feed through the constraint system all the constraints
implied by a concrete type satisfying a conformance requirement.

huonw committed:
[Sema] Add two FIXMEs.

huonw commented:
@swift-ci please smoke test

huonw commented on test/Generics/conditional_conformances.swift:
```diff
@@ -0,0 +1,204 @@
+// CHECK-NEXT: (normal_conformance type=SameType<Int> protocol=P2
+// CHECK-NEXT:   same_type: τ_0_0 Int)
+extension SameType: P2 where T == Int {}
+// FIXME: the compiler gets this... exactly backwards. :( For the incorrectly
```
@DougGregor This test is fun. It looks like with the `T == Int` the `SameType<T>` is eagerly "resolved" to `SameType<Int>` in conformances, meaning `SameType<Int>` isn't treated as a specialised conformance, and that's generally confusing the compiler.

huonw commented:
This now handles most of both the decl checking and the (type-level) expr checking parts of conditional conformances. It still doesn't touch witness tables, but I'll probably do that in a follow up.

huonw commented:
@swift-ci please smoke test.

DougGregor commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -285,6 +287,132 @@ AbstractStorageDecl *ProtocolConformance::getBehaviorDecl() const {
+  // every thing in the type signature either is included too or is implied by
+  // something else. The most important bit is having the same type
+  // parameters. (NB. if/when Swift gets parameterised extensions, this needs to
+  // change to the extension's params being a superset of the type's.)
```
... and now that we've discussed this in person, I see what you mean and your representation could definitely be simpler than what I was proposing :)

DougGregor commented on lib/Sema/CSApply.cpp:
```diff
@@ -2232,6 +2240,8 @@ namespace {
                               ConformanceCheckFlags::InExpression);
       assert(conformance && "object literal type conforms to protocol");
+      assert(conformance->getConditionalRequirements().empty() &&
+             "unhandled conditional conformance");
```
... doesn't have to be in the first PR, of course!

huonw replied:
Yeah, definitely; just wanted to establish some "ground rules"/cover the compiler in asserts to flag the places I'll needed to update when compiling a simple example.

DougGregor commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -1257,6 +1257,10 @@ RequirementCheckResult TypeChecker::checkGenericArguments(
       case RequirementCheckResult::Success:
+        // FIXME: we should possibly have a queue of requirements (or
+        // ArrayRef<Requirements>) that this function works through, and add the
+        // conditional requirements to the end of this.
```
Yeah, that makes sense. You're discovering additional requirements along the way that'll need to be checked. The "listener" interface might be a little tricky.

DougGregor reviewed (commented):
This is coming along wonderfully! The changes are numerous, but they're all quite straightforward uses of the new "conditional requirements" for the protocol.

huonw commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -285,6 +287,132 @@ AbstractStorageDecl *ProtocolConformance::getBehaviorDecl() const {
+/// Foo or T == U (etc.) with the information it knows. This includes checking
+/// against global state, if any/all of the types in the requirement are
+/// concrete, not type parameters.
+static bool isRequirementSatisfied(GenericSignature *sig,
```
I actually moved it out of the GenericSignature on the advice of @DougGregor (you should argue with that guy 😛). In particular, this will be checking global conformances/using the module, but it sounds like GenericSignature wants to remove its connection to that. Maybe it could go on the module decl, or just `swift::isRequirementSatisfied` top-level somewhere?

DougGregor replied:
The new `GenericSignatureBuilder::lookupConformance`  should be used to answer these queries.

DougGregor replied:
(Future-me still wants the API on `GenericSignature` :))

huonw replied:
Done.

huonw commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2613,6 +2613,16 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyConformsToConstraint(
+      // phrased in terms of the type's generic parameters, not type variables
+      // in this context.
+      for (auto req : conformance->getConditionalRequirements()) {
+        addConstraint(req, locator);
```
Unsure. I think I might've noticed something strange, and haven't fully diagnosed yet.

huonw commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -285,6 +287,132 @@ AbstractStorageDecl *ProtocolConformance::getBehaviorDecl() const {
+    // Walk up the chain of classes looking for the one in the superclass
+    // bound. We have to make sure we're considering things in this context,
+    // because type parameters in superclasses could be equated.
+    while (1) {
```
Oh, right, I can pull out the superclass for the first step and then use that. Thanks.

huonw commented on include/swift/AST/ProtocolConformance.h:
```diff
@@ -365,6 +373,7 @@ class NormalProtocolConformance : public ProtocolConformance,
     : ProtocolConformance(ProtocolConformanceKind::Normal, conformingType),
       ProtocolAndState(protocol, state), Loc(loc), ContextAndInvalid(dc, false)
   {
+    differenceAndStoreConditionalRequirements();
```
Oh, so store it in the `ExtensionDecl` or something? I'm not sure where else this info should be stored.

DougGregor replied:
Or store it in the `NormalProtocolConformance`. Those are serialized, too.

huonw replied:
I haven't done this yet, but I will in a follow up: since the stdlib (etc.) isn't using this yet, it isn't necessary and I/you want this to be merged.

huonw committed:
[GSB] subst Requirements directly; no need for the extra argument.

huonw committed:
[Sema] Check conditional conformances for generic function calls early.

This allows for better (although not perfect) diagnostics, and more guarantees
of correctness.

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please smoke test and merge.
